lib/ratfun.gi

```diff
@@ -1304,18 +1304,18 @@ end);
 ##  the methods are implemented.
 ##  
 ##
-InstallMethod( GcdOp,"Gcd(Polyring, Pol,Pol)",
-    IsCollsElmsElms,[IsEuclideanRing,
-                IsRationalFunction,IsRationalFunction],0,
-function(R,f,g)
-
-	if IsUnivariatePolynomial(f) and IsUnivariatePolynomial(g) 
-		and IndeterminateNumberOfUnivariateRationalFunction(f) = 
-		IndeterminateNumberOfUnivariateRationalFunction(g) then
-		return GcdOp(R,f,g);
-	fi;
+InstallRingAgnosticGcdMethod("test polynomials for univar. and same variable",
+  IsCollsElmsElms,IsIdenticalObj,
+  [IsEuclideanRing,IsRationalFunction,IsRationalFunction],0,
+function(f,g)
+
+  if IsUnivariatePolynomial(f) and IsUnivariatePolynomial(g) 
+    and IndeterminateNumberOfUnivariateRationalFunction(f) = 
+    IndeterminateNumberOfUnivariateRationalFunction(g) then
+    return GcdOp(f,g);
```

Comment:
Polynomials (well, rational functions) do not automatically default into a "most compact" representation, as testing is does not come for free, and there frankly (in longer arithmetic) there is no value to briefly drop out of the more general representation only for the next arithmetic step requires to jump back into the same general representation. Thus an univariate rational function might not know yet that it is, if it arises somewhere out of a calculation, and the user might not have thought about explicitly enforcing an univariateness test. (This was indeed an issue in a prior version of the polynomials code.)

There is already a GcdOp method "A" for (known) univariate polynomials.

This method "B" is an enforced fallback test (to avoid going into Groebner basis land for a multivariate test). Because the requirements are lower, it will rank below method "A". This means that the polynomials could not have known to be univariate when the function was called. That is the "learned" information.
(Well, the test for same indeterminate was added later and makes the argument slightly wrong...)

For the last 10 or 15 years I have written such conditions more carefully by adding the requirement that one of the polynomilas does not know its univariate status. I will also do so here.